src/puppetlabs/puppetdb/command/dlo.clj

```diff
@@ -1,210 +1,144 @@
-                                     :let [files (compressible-files subdir threshold)]
-                                     :when (seq files)]
-                                 [subdir files]))]
-    (time! (global-metric :compression)
-           (doseq [[subdir files] subdir-files]
-             (compress-subdir! subdir files)))))
+  (:require
+   [clojure.java.io :as io]
+   [clojure.string :as str]
+   [metrics.counters :as counters]
+   [puppetlabs.i18n.core :refer [trs]]
+   [puppetlabs.kitchensink.core :refer [timestamp]]
+   [puppetlabs.puppetdb.nio :refer [copts copt-atomic copt-replace]]
+   [puppetlabs.puppetdb.queue
+    :refer [cmdref->entry metadata-cmd-names parse-metadata]]
+   [stockpile :as stock])
+  (:import
+   [java.nio.file AtomicMoveNotSupportedException
+    FileAlreadyExistsException Files LinkOption]
+   [java.nio.file.attribute FileAttribute]))
+
+(defn- parse-cmd-filename [s]
+  (let [rx #"([0-9+])-(.*)"]
+    (when-let [[_ id qmeta] (re-matches rx s)]
+      (parse-metadata qmeta))))
+
+(defn- cmd-counters
+  "Adds gauges to the dlo for the given category (e.g. \"all\"
+  \"replace command\").  The gauges will pull their values from the
+  dlo atom's :stats."
+  [registry category]
+  (let [dlo-ns "puppetlabs.puppetdb.dlo"]
+    {:count (counters/counter registry [dlo-ns category "count"])
+     :size (counters/counter registry [dlo-ns category "size"])}))
+
+(defn- update-metrics
+  "Updates stats to reflect the receipt of the named command."
+  [metrics command size]
+  (let [all (metrics "all")
+        cmd (metrics command)]
+    (counters/inc! (:count all))
+    (counters/inc! (:size all) size)
+    (counters/inc! (:count cmd))
+    (counters/inc! (:size cmd) size)))
+
+(defn- update-metrics-from-path
+  "Updates metrics to reflect the discarded commands directly inside
+  the path; does not recurse."
+  [metrics path]
+  (with-open [path-stream (Files/newDirectoryStream path)]
+    (doseq [path path-stream]
+      ;; Assume the trailing .json here and in entry-cmd-err-filename below.
+      (when-let [cmd (:command
+                      (and (.endsWith path ".json")
+                           (parse-cmd-filename (-> path .getFileName str))))]
+        (update-metrics metrics cmd (Files/size path))))))
+
+(defn- write-failure-metadata
+  "Given a (possibly empty) sequence of command attempts and an exception,
+  writes a summary of the failure to out."
+  [out attempts exception]
+  (let [write-header (fn [out i timestamp]
+                       (.format out "%sAttempt %d @ %s\n"
+                                (into-array Object [(if (zero? i) "" "\n")
+                                                    (inc i)
+                                                    timestamp])))]
+    (dorun (map-indexed (fn [i {:keys [timestamp error trace]}]
+                          (write-header out i timestamp)
+                          (.print out trace))
```

Comment:
Attempts aren't exceptions, they're just serialized backtrace (preexisting behavior), though I've tweaked them in this pr to just be the one big printStackTrace string rather than a list of stacktrace "elements" -- an intermediate step; the forthcoming pr (already up for review) will revamp this to make everything an exception.